User can select who goes first

lpenzey opened the pull request:
User is now prompted to select which player goes first. They must enter
the exact player/opponent token to proceed, otherwise they're shown an
error message and are prompted again to choose which player they would
like to go first. Their choice is then set in the current_player field
in the State struct.

The functionality is implemented in Game.play, where the turn is
determined by who the current player is as determined by querying
state.current_player.

lpenzey committed:
User can select who goes first

User is now prompted to select which player goes first. They must enter
the exact player/opponent token to proceed, otherwise they're shown an
error message and are prompted again to choose which player they would
like to go first. Their choice is then set in the current_player field
in the State struct.

The functionality is implemented in Game.play, where the turn is
determined by who the current player is as determined by querying
state.current_player.

ssunday commented on lib/TTT/validation.ex:
```diff
@@ -39,6 +39,23 @@ defmodule Validation do
+      Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
+    )
+    |> clean()
+    |> validate_first_player(io, player_token, opponent_token)
```
Would it make more sense to throw an error in validate, then have this case off of it and recurse within itself?

lpenzey replied:
like this?

``` 
  def choose_first_player(io, player_token, opponent_token) do
    first_player =
      io.retrieve(
        Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
      )
      |> clean()

    case validate_first_player(first_player, player_token, opponent_token) do
      {:ok} ->
        first_player

      {:error} ->
        io.display(Messages.get_message(:first_player_error))
        choose_first_player(io, player_token, opponent_token)
    end
  end

  def validate_first_player(first_player, player_token, opponent_token) do
    case first_player do
      ^player_token ->
        {:ok}

      ^opponent_token ->
        {:ok}

      _ ->
        {:error}
    end
  end
```

ssunday replied:
Yeah, or piping validate_first_player to `|> clean()`

```
def choose_first_player(io, player_token, opponent_token) do
    io.retrieve(
        Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
      )
      |> clean()
      |> validate_first_player(player_token, opponent_token)
      |> case do
      {:ok} ->
        first_player

      {:error} ->
        io.display(Messages.get_message(:first_player_error))
        choose_first_player(io, player_token, opponent_token)
    end
  end

  def validate_first_player(first_player, player_token, opponent_token) do
    case first_player do
      ^player_token ->
        {:ok}

      ^opponent_token ->
        {:ok}

      _ ->
        {:error}
    end
  end
```

lpenzey committed:
Refactor first_player validation

rmulhol commented:
How's the performance if the game is setup with the unbeatable computer player going first?

rmulhol commented on lib/TTT/game.ex:
```diff
@@ -5,32 +5,40 @@ defmodule Game do
+        |> check_status(io, mode)
+
+      ^opponent ->
+        opponent(state, io, mode)
```
This is pretty minor but I'd advise against overloading an overloaded function with a variable of the same name

lpenzey commented:
> How's the performance if the game is setup with the unbeatable computer player going first?

Quite slow (~24 seconds). Easy fix is to hard-code the first move to be in one of the corners, but I was trying to tackle it using processes to speed things up.

lpenzey committed:
Refactor opponent functions for clarity

lpenzey committed:
Update modules to use state getters

Added a getter in State for the current player, and then updated all
calls that access state internals directly and rerouted them through the
getter functions.
